app/scripts/lib/AddressBookPetnamesBridge.ts

```diff
@@ -0,0 +1,247 @@
+import {
+  AddressBookController,
+  AddressBookState,
+} from '@metamask/address-book-controller';
+import { RestrictedControllerMessenger } from '@metamask/base-controller';
+import {
+  NameController,
+  NameControllerState,
+  NameStateChange,
+  NameType,
+} from '@metamask/name-controller';
+import { cloneDeep } from 'lodash';
+import log from 'loglevel';
+
+type Entry = {
+  address: string;
+  name: string;
+  chainId: string;
+  isEns: boolean;
+};
+
+type AllowedEvents = NameStateChange;
+
+export type AddressBookPetnamesBridgeMessenger = RestrictedControllerMessenger<
+  'AddressBookPetnamesBridge',
+  never,
+  AllowedEvents,
+  never,
+  AllowedEvents['type']
+>;
+
+export class AddressBookPetnamesBridge {
+  #addressBookController: AddressBookController;
+
+  #addressBookState: AddressBookState;
+
+  #nameController: NameController;
+
+  #nameState: NameControllerState;
+
+  #messenger: AddressBookPetnamesBridgeMessenger;
+
+  #updating: boolean;
+
+  constructor({
+    addressBookController,
+    nameController,
+    messenger,
+  }: {
+    addressBookController: AddressBookController;
+    nameController: NameController;
+    messenger: AddressBookPetnamesBridgeMessenger;
+  }) {
+    this.#addressBookController = addressBookController;
+    this.#addressBookState = addressBookController.state;
+    this.#nameController = nameController;
+    this.#nameState = nameController.state;
+    this.#messenger = messenger;
+    this.#updating = false;
+  }
+
+  init() {
```

Comment:
Shouldn't we at least trigger once on `init`? Otherwise we need to wait atleast one change per to sync, or do I miss something?

Reply:
We have the migration for any historic address book entries, and we're not concerned about existing petnames entries as that would be dev builds only.

The bridge then catches any new changes on either side.

Reply:
But migration will be shipped all users, while `petnames` shipped flask only. If normal users keep using Addressbook before getting feature, there will be a gap then. 
It will not be rare case but will be resolved eventually anyway.

Reply:
You're absolutely correct, but I don't believe we support build type specific migrations, so I was assuming we'd duplicate the migration before the main release to cover the latest address book entries for the main users.